tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -5,149 +5,190 @@
+ *         really been written (the same way that {@link java.io.RandomAccessFile#write(byte[])}
+ *         would write them).</li>
+ *     <li>{@link #invalidateBufferedData()} ensures that all previously-performed
+ *         writes through other file descriptors are now readable through this one.</li>
+ * </ul>
+ *
+ * @author Allan Heydon
  */
-
-import java.io.File;
-import java.io.IOException;
-import java.math.BigInteger;
-
 public final class BufferedRandomAccessFile extends java.io.RandomAccessFile {
-	// Increase 8k buffer to match modern day hardware? No!!! The implementation
-	// discards the whole buffer each time it seeks to a position outside the
-	// current buffer. It then creates a new buffer, which takes proportionally
-	// longer with larger buffer sizes.
-	static final int LogBuffSz = 13; // 8K buffer
+    // Increase 8k buffer to match modern day hardware? No!!! The implementation
+    // discards the whole buffer each time it seeks to a position outside the
+    // current buffer. It then creates a new buffer, which takes proportionally
+    // longer with larger buffer sizes.
+    static final int LogBuffSz = 13; // 8K buffer
     public static final int BuffSz = (1 << LogBuffSz);
     static final int BuffMask = ~(BuffSz - 1);
 
     /* This implementation is based on the buffer implementation in
-       Modula-3's "Rd", "Wr", "RdClass", and "WrClass" interfaces. */
+       Modula-3's "Rd", "Wr", "RdClass", and "WrClass" interfaces,
+       with substantial modifications. */
+
+    /**
+     * True if unflushed bytes exist.
+     * @see #flushBuffer()
+     */
     private boolean dirty;  // true iff unflushed bytes exist
+
     // SZ Feb 24, 2009: never read
     // private boolean closed; // true iff the file is closed
-    private long curr;      // current position in file
-    private long lo, hi;    // bounds on characters in "buff"
-    private byte[] buff;    // local buffer
-    private long maxHi;     // this.lo + this.buff.length
-    private boolean hitEOF; // buffer contains last file block?
-    private long diskPos;   // disk position
-	private long mark;
-    
-    private static final Object mu = new Object(); // protects the following fields
+
+    /** Current length of the file. */
+    private long length;
+
+    /** Current position in the file. */
+    private long curr;
+
+    /** Offset of <code>buff[0]</code> in the file. */
+    private long lo;
+
+    /** Cache of the bytes at indexes <code>[lo, min(lo + BuffSz, length))</code>. */
+    private final byte[] buff;
+
+    /**
+     * Cached value of <code>super.{@link #getFilePointer()}</code>.  Used to avoid
+     * some calls to <code>super.{@link #seek(long)}</code>.
+     */
+    private long diskPos;
+
+    /**
+     * Value of the mark.
+     * @see #mark()
+     * @see #reset()
+     */
+    private long mark;
+
+    /** Protects {@link #availBuffs} and {@link #numAvailBuffs}. */
```

Comment:
`mu` also seems to protect `buff`.

Reply:
Not quite; `availBuffs` and `numAvailBuffs` implement a shared [free list](https://en.wikipedia.org/wiki/Free_list) for allocating buffers. Once this class has allocated a buffer by removing one from `availBuffs`, it does not need to lock anything to access its privately allocated `buff`.

More details in this comment --> https://github.com/tlaplus/tlaplus/pull/907#discussion_r1573045418